api/openapi/model-registry.yaml

```
@@ -1244,17 +1205,6 @@ components:
                 since epoch.
               type: string
               readOnly: true
```

While it seems like a good idea, readOnly is only for documentation:
```
JSON Schema's readOnly field is primarily for documentation and API contract purposes and is not inherently enforced by Go's encoding/json package during serialization or deserialization. While the readOnly keyword can be included in a JSON schema definition, Go will not automatically prevent modification or ignore the field. Handling readOnly fields requires custom logic.
```
That's why the generated code has no enforcement or safeguard anywhere. We had investigated this in RedHat managed services and discovered this issue a few years ago.

While it seems like a good idea, users are going to completely miss the readOnly _tag_ and try to send those properties. Service code will also have to manually hand written to ignore/handle them appropriately. What should the service do when readonly fields are sent by the client when they shouldn't, should it throw an error/ignore them? 

This will make the API ambiguous and the implementation complicated. The current separate types schema gives us a fool proof separation of concerns, and simplifies our service implementation which is able to leverage generation of goverter helpers that automatically filter fields for us.

---

Thank you @dhirajsb for the blurb from the OpenAPI spec. I did not know this before.

I agree that `readOnly` does not provide strict enforcement. However, I respectfully disagree with the concern that this change makes the API ambiguous or complicated. On the contrary, it improves both the API’s readability and the experience of code generation and simplicity in extending the API.

Given that clients are accidentally sending readOnly properties, it indicates that API clients are not adhering to the API specification as intended. The enforcement behavior on the backend is an implementation detail, and I don't see a strong justification to complicate the API design because of it.

On the backend, we should explore
1. Silently ignore them if they were present in a request and/or
2. Find a way to generate the enforcement through our extensions for the code generation

While I appreciate the enforcement provided by the current approach, I also feel the API design is too closely tied to the domain model. IMO, prioritizing overall readability and making the API easier to extend in the future would add more long-term value.

---

there's the possibility of using "not" , as a way to exclude a field from a verb like enable "name" in create and remove it in update

sources:

https://github.com/OAI/OpenAPI-Specification/issues/1497#issuecomment-370951318

https://github.com/OAI/OpenAPI-Specification/issues/1497#issuecomment-371249646

---

The current version generates the server side code to ignore the `readOnly` values when they are encountered in the `POST` and `PATCH` requests, thus keeping existing functionality as is.